Installing OpenMC using Docker

wbinventor opened the pull request:
This PR adds a new [`Dockerfile`](https://docs.docker.com/engine/reference/builder/) to make it trivial to deploy OpenMC on any Linux, Mac or Windows machine running [Docker](https://www.docker.com/). The quick install and installation instructions in the user's guide have been updated with a summary on how to build a Docker image on one's local machine with OpenMC fully installed (all tests pass), and how to spawn one or more containers from that image to run OpenMC.

wbinventor commented on Dockerfile:
```diff
@@ -0,0 +1,47 @@
+    apt-get autoremove
+
+# Download Miniconda3 and install Python dependencies
+RUN wget --quiet https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -O ~/miniconda.sh && \
```
Fine by me - I just chose conda since I wanted to have it installed on the image I'll be doing development from, but I can install it separately.

paulromano replied:
Ah, you're actually using it :) My primary concern was just regarding the size of the image. Do you know if it will make a difference whether we use conda or the system-provided Python?

wbinventor replied:
First, I like the idea of using the system-provided Python to simplify the `Dockerfile` and to more closely mirror the installation instructions. As for image size, about 240MB is saved by dropping Miniconda: the Miniconda image is 5.21GB while the system image is 4.97GB (2.5GB consumed by the nuclear data). I'll probably continue to use Miniconda in my own personal image for development, but fortunately Docker makes it simple to manage multiple images :-)

wbinventor commented on Dockerfile:
```diff
@@ -0,0 +1,47 @@
+    apt-get install -y wget git emacs && \
+    apt-get install -y gfortran g++ cmake && \
+    apt-get install -y mpich libmpich-dev && \
+    apt-get install -y libhdf5-serial-dev libhdf5-mpich-dev && \
```
Done

wbinventor commented on Dockerfile:
```diff
@@ -0,0 +1,47 @@
+    /bin/bash ~/miniconda.sh -b -p /opt/conda && \
+    rm -rf ~/miniconda.sh
+RUN pip install --upgrade pip && \
+    pip install six numpy scipy pandas h5py matplotlib uncertainties lxml cython jupyter pytest && \
```
Done

wbinventor commented on Dockerfile:
```diff
@@ -0,0 +1,47 @@
+RUN mkdir /opt/openmc/data
+RUN wget https://anl.box.com/shared/static/a0eflty17atnpd0pp7460exagr3nuhm7.xz -O - | tar -C /opt/openmc/data -xvJ
+RUN wget https://anl.box.com/shared/static/4kd2gxnf4gtk4w1c8eua5fsua22kvgjb.xz -O - | tar -C /opt/openmc/data -xvJ
+RUN wget https://github.com/mit-crpg/WMP_Library/releases/download/v1.0/WMP_Library_v1.0.tar.gz && \
```
I tried this but ran into trouble getting `xvfb` working as called in [`travis-before-script.sh`](https://github.com/wbinventor/openmc/blob/develop/tools/ci/travis-before-script.sh#L6). Although there are quite a lot of stackoverflow posts and gists describing ways to fix this in Docker containers, I was unsuccessful. I know it would introduce redundancy, but if we care about simplifying the `Dockerfile` further, perhaps we could call a new `docker-download-data.sh` that would mirror `travis-before-script.sh` but would not start `xvfb`?

wbinventor commented on docs/source/quickinstall.rst:
```diff
@@ -58,6 +58,49 @@ are no longer supported.
+Installing with Docker on Linux/Mac/Windows
+-------------------------------------------
+
+OpenMC can be easily deployed using `Docker <https://www.docker.com/>`_ on any
```
Sure, this sounds fine to me. I can simplify the quick install and move this section to the developer's guide. I just thought it might make sense to include this in the installation instructions since I would argue that this may now be the simplest way for new users -- and primarily students -- to painlessly "install" OpenMC on whatever platform they may be using without running into dependency hell. But of course Docker has its own learning curve, and thus it might be of most interest to "power users" and developers.

wbinventor replied:
Btw, if you think it would be a good idea to refer users to Dockerhub in the quick install guide, what do you think about creating an `openmc-dev` Dockerhub user with an `openmc` repo?

paulromano replied:
Good idea. I just created an `openmc` organization and added you as an owner.

wbinventor replied:
My latest commit simplified the quick installation by referring user's to `run` the image from DockerHub. I moved the more detailed description of how to build a Docker image to a new sub-section at the end of the Dev Guide.

wbinventor commented:
@paulromano to be honest I hadn't thought much about how often I would update the Docker image on Dockerhub. I just thought I'd push the image there to test out the platform and show an example of how an image can be pulled for this PR. I do agree that if we want to manage images on Dockerhub, it would probably make the most sense to just build and push them for each new releases.

paulromano commented on docker/docker-download-xs.sh:
```diff
@@ -0,0 +1,19 @@
+
+# Download NNDC HDF5 data
+if [[ ! -e $HOME/nndc_hdf5/cross_sections.xml ]]; then
+    wget https://anl.box.com/shared/static/a0eflty17atnpd0pp7460exagr3nuhm7.xz -O - | tar -C $HOME -xvJ
```
I still have the same objection in that these URLs appear in two different files. Maybe we should just call this one `tools/ci/download-xs.sh` and then call it from `tools/ci/travis-before-script.sh` (as well as from the Dockerfile)

wbinventor commented:
@paulromano I've addressed all of your comments, but have not yet pushed an image for v0.10.0 to DockerHub. I'll do so when I have a good internet connection again (the Comcast curse has struck my apt again). It's up to you whether to merge this before or after an image has been pushed to the Hub.

paulromano commented on docs/source/quickinstall.rst:
```diff
@@ -58,6 +58,30 @@ are no longer supported.
+This will take 5 -- 10 minutes to run to completion depending on your internet download speed. The command will place you in an interactive shell running in a `Docker container`_ with OpenMC installed.
+
+.. note:: The ``docker run`` command supports many
+          `options <https://docs.docker.com/engine/reference/commandline/run/>`_
```
``` `options`_ ```

paulromano commented:
I like the new organization of the documentation, thanks! Since there's no rush, I'll wait until the new image is up on Docker hub to merge.

wbinventor commented:
I pushed a Docker image of v0.10.0 to the Dockerhub ([link](https://hub.docker.com/r/openmc/openmc/)). This can be pulled and run with the following:


```Shell
> docker run openmc/openmc:v0.10.0
```

I had to revise the `Dockerfile` somewhat for v0.10.0 since the dev team has been so active in the last 9 months 😉. Let me know if you see anything else you'd like revised before merging @paulromano.
